src/renderers/WebGLRenderer.js

```diff
@@ -49,6 +49,10 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	this.sortObjects = true;
 
+	// user-defined clipping
+
+	this.userClippingPlanes = [];
```

Comment:
I left the planes array on the renderer for now (renamed to `.clippingPlanes`). I could fairly easily put the interface on the main camera, but I'm still undecided.

> There are several use cases where it would be cool to allow it on a per object basis. I don't think it really belongs onto the material, but it would be the "reasonably close approximation right now" to avoid issues like #8379 and keep this PR focussed (I guess that's why you are suggesting it?). 

I guess, I might have guessed wrong. Per-material planes would allow clipping a whole group of objects at once...

Now that the update paths are all working can shape it. In a while - I need to read over it again with a clear mind, maybe clean up a bit and think out how to best implement per-material clipping.